NeosModLoader/Utility/GenericTypeMethodsInvoker.cs

```diff
@@ -0,0 +1,166 @@
+using HarmonyLib;
+using NeosModLoader;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NeosModLoader.Utility
+{
+	public sealed class GenericTypeMethodsInvoker
+	{
+		private readonly Dictionary<TypeDefinition, ConcreteType> concreteTypes = new();
+
+		public Type GenericType { get; }
+
+		public Func<MethodInfo, Type, MethodInfo> GetGenericMethodOfConcreteType { get; }
+
+		public GenericTypeMethodsInvoker(Type genericType)
+			: this(genericType, GetGenericMethodOfConcreteTypeDefault)
+		{ }
+
+		public GenericTypeMethodsInvoker(Type genericType, Func<MethodInfo, Type, MethodInfo> getGenericMethodOfConcreteType)
+		{
+			GenericType = genericType;
+			GetGenericMethodOfConcreteType = getGenericMethodOfConcreteType;
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type[] instanceTypes, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceTypes, instance, methodTypes, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type[] instanceTypes, object instance, Type methodType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
```

Comment:
... I've not written C# in quite a while, but to me it seems a bit bad/weird to have 8 overwrite methods calling the same "internal" method with the same parameters.... Since like, if the internal method can seemingly accept all of the different parameter signatures just fine, why couldn't the external one too..?

Reply:
On top of the ambiguous match mentioned in the other comment, this also hides the internally used TypeDefinition type from the public interface.

Reply:
Yea, may be worth it - would cut it more than in half with the generic / non generic and instance parameter versions stay around